File: ruby/SpaceStation.rb
#encoding-utf:8

require_relative 'SpaceStationToUI'
require_relative 'CardDealer'
require_relative 'Transformation'

module Deepspace
   class SpaceStation
        @@MAXFUEL = 100             #maxima cantidad de combustible
        @@SHIELDLOSSPERUNITSHOT = 0.1   #unidades de escudo que se pierden por unidad de potencia de disparo recibido
        def initialize(n, s) #string, suppliespackage
            @name = n
            @ammoPower = s.ammoPower           #float
            @nMedals = 0                        #int
            @fuelUnits =  s.fuelUnits          #float
            @shieldPower = s.shieldPower         #float
            @pendingDamage = nil
            @hangar = nil
            @weapons  = Array.new
            @shieldBoosters = Array.new
        end

        private
        def assignFuelValue(f)
            if ( f > @@MAXFUEL)
                @fuelUnits = @@MAXFUEL
            else
                @fuelUnits = f
            end
        end

        def cleanPendingDamage
            if( @pendingDamage.hasNoEffect )
                @pendingDamage = nil
            end
        end

        public
        def cleanUpMountedItems #elimina elementos con 0 usos
            wnu = []
            snu = []
            @weapons.each{ |w|
                if w.uses == 0
                    wnu.push(w)
                end
            }
            wnu.each{ |wNoUses|
                @weapons.delete(wNoUses)
            }
            @shieldBoosters.each { |sb|
                if sb.uses == 0
                    snu.push(sb)
                end
            }
            snu.each{ |sbNoUses|
                @shieldBoosters.delete(sbNoUses)
            }
        end

        def copy(s)
            @ammoPower = s.ammoPower
            @name = s.name
            @shieldPower = s.shieldPower
            @fuelUnits = s.fuelUnits
            @nMedals = s.nMedals

            @pendingDamage = s.pendingDamage
            @weapons = s.weapons
            @shieldBoosters = s.shieldBoosters
            @hangar = s.hangar
        end

        def discardHangar
            @hangar = nil
        end

        def discardShieldBooster(i) #i int
            size = @shieldBoosters.length
            if ( i>=0 && i<size)
                s = @shieldBoosters.delete_at(i)
                if @pendingDamage != nil
                    @pendingDamage.discardShieldBooster
                    cleanPendingDamage
                end
            end
        end

        def discardShieldBoosterInHangar(i) #i int
            if (@hangar != nil)
                @hangar.removeShieldBooster(i)
            end
        end

        def discardWeapon(i) #i int
            size = @weapons.length
            if ( i>=0 && i<size)
                w = @weapons.delete_at(i)
                if @pendingDamage != nil
                    @pendingDamage.discardWeapon(w)
                    cleanPendingDamage
                end
            end
        end

        def discardWeaponInHangar(i)
            if (@hangar != nil)
                @hangar.removeWeapon(i)
            end
        end

        def fire
            size = @weapons.length
            factor = 1.0

            for i in (0..size-1)
                w = @weapons.at(i)
                factor *= w.useIt
            end

            return ammoPower*factor
        end

        attr_reader :ammoPower
        attr_reader :fuelUnits
        attr_reader :hangar
        attr_reader :name
        attr_reader :nMedals
        attr_reader :pendingDamage
        attr_reader :shieldBoosters
        attr_reader :weapons
        attr_reader :shieldPower

        def speed
            return @fuelUnits.to_f/@@MAXFUEL.to_f
        end

        def getUIversion
            return SpaceStationToUI.new(self)
        end

        def mountShieldBooster(i)
            if (@hangar != nil && i>=0 && i<@hangar.shieldBoosters.length)
                sb = @hangar.removeShieldBooster(i)
                if(sb != nil)
                  @shieldBoosters.push(sb)
                end
            end
        end

        def mountWeapon(i)
            if (@hangar != nil && i>=0 && i<@hangar.weapons.length)
                #weapon = @hangar.weapons[i]
                weapon = @hangar.removeWeapon(i)
                if(weapon != nil)
                  @weapons.push(weapon)
                end
            end
        end

        def move
            @fuelUnits -= (speed*fuelUnits)
            if @fuelUnits<0
                @fuelUnits = 0
            end
        end

        def protection
            size = @shieldBoosters.length
            factor = 1.0

            for i in (0..size-1)
                s = @shieldBoosters.at(i)
                factor *= s.useIt
            end
            
            return shieldPower*factor
        end

        def receiveHangar(h)
            if (@hangar == nil)
                @hangar = h
            end
        end

        def receiveShieldBooster(s)
            if (@hangar != nil)
                return @hangar.addShieldBooster(s)
              else
                false
            end
        end

        def receiveShot(shot) #shot es float
            myProtection = self.protection
            if(myProtection>=shot)
                @shieldPower -= @@SHIELDLOSSPERUNITSHOT * shot
                @shieldPower = [0.0,@shieldPower].max
                return Deepspace::ShotResult::RESIST
            else
                @shieldPower = 0.0
                return Deepspace::ShotResult::DONOTRESIST
            end

        end

        def receiveSupplies(s)
            @ammoPower += s.ammoPower
            @fuelUnits += s.fuelUnits
            @shieldPower += s.shieldPower

            if(@fuelUnits > @@MAXFUEL)
                @fuelUnits = @@MAXFUEL
            end
        end

        def receiveWeapon(w)
            if(hangar != nil)
                return @hangar.addWeapon(w)
              else return false
            end
        end

        def setLoot(loot)
            dealer = Deepspace::CardDealer.instance
            h = loot.nHangars

            if(h>0)
                hangar = dealer.nextHangar()
                receiveHangar(hangar)
            end

            elements = loot.nSupplies
            for i in (1..elements)
                sup = dealer.nextSuppliesPackage
                receiveSupplies(sup)
            end

            elements = loot.nWeapons
            for i in (1..elements)
                weap = dealer.nextWeapon
                receiveWeapon(weap)
            end

            elements = loot.nShields
            for i in (1..elements)
                sh = dealer.nextShieldBooster
                receiveShieldBooster(sh)
            end

            medals = loot.nMedals
            @nMedals += medals

            if loot.efficient
                return Deepspace::Transformation::GETEFFICIENT
            end 
            if loot.spaceCity
                return Deepspace::Transformation::SPACECITY
            end
    
            return Deepspace::Transformation::NOTRANSFORM
        end

        def setPendingDamage(d)
            @pendingDamage = d.adjust(@weapons, @shieldBoosters)
            cleanPendingDamage
        end

        def validState
            if @pendingDamage == nil
                return true
            end

            return @pendingDamage.hasNoEffect
        end

        def to_s
            pd = ", Pending Damage: "
            if pendingDamage != nil
                pd += "#{@pendingDamage.to_s}"
            else    pd += "nil"
            end

            we = ", Weapons mounted: "
            if (!@weapons.empty?)
                @weapons.each { |aux|
                    we += "#{aux.to_s}"
                }
            end

            sh = ", Shield Boosters mounted: "
            if (!@shieldBoosters.empty?)
                @shieldBoosters.each { |aux|
                    sh += "#{aux.to_s}"
                }
            end

            h = "Hangar:"
            if hangar != nil
                h += "#{@hangar.to_s}"
            else    h += "nil"
            end

            return "Name: #{@name}, Ammo Power: #{@ammoPower}, Fuel Units: #{@fuelUnits}, Shield Power: #{@shieldPower}, NMedals: #{@nMedals}" + pd + we + sh + h

        end
   end 
end

File: java/Deepspace/src/deepspace/SpaceStation.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;
import java.util.ArrayList;
/**
 *
 * @author alvaro
 */
public class SpaceStation implements SpaceFighter{
    private static final float MAXFUEL = 100f;
    private static final float SHIELDLOSSPERUNITSHOT = 0.1f; 
    
    private float ammoPower;
    private float fuelUnits;
    private String name;
    private int nMedals;
    private float shieldPower;
    private Damage pendingDamage;
    private ArrayList<Weapon> weapons = new ArrayList<>();
    private ArrayList<ShieldBooster> shieldBoosters = new ArrayList<>();
    private Hangar hangar;
    
    SpaceStation(String n, SuppliesPackage supplies){
        name = n;
        ammoPower = supplies.getAmmoPower();
        fuelUnits = supplies.getFuelUnits();
        shieldPower = supplies.getShieldPower();
        nMedals = 0;
        pendingDamage = null;
        /*weapons = null;
        shieldBoosters = null;      //si pongo estos nulls peta al iniciar
        hangar = null;*/
    }
    
    SpaceStation(SpaceStation station){
        this(station.getName(), new SuppliesPackage(station.getAmmoPower(), station.getFuelUnits(), station.getShieldPower()));
        pendingDamage = station.pendingDamage;
        nMedals = station.nMedals;
        
        station.weapons.forEach(w -> {
            weapons.add(w);
        });
        station.shieldBoosters.forEach(s -> {
            shieldBoosters.add(s);
        });
        hangar = station.hangar;
    }
    
    private void assignFuelValue(float f){
        if (f < MAXFUEL) {
            this.fuelUnits = f;
        } else {
            this.fuelUnits = MAXFUEL;
        }
    }
    
    private void cleanPendingDamage(){
        if(pendingDamage.hasNoEffect()){
           pendingDamage = null;             
        }
    }
    
    public void cleanUpMountedItems(){
        ArrayList<Weapon> auxweapons = new ArrayList<>();
        ArrayList<ShieldBooster> auxshield = new ArrayList<>();
        for(Weapon w: weapons){
            if (w.getUses() == 0)
                auxweapons.add(w);
        }
        for(Weapon w: auxweapons){
            weapons.remove(w);
        }
        
        for(ShieldBooster s: shieldBoosters){
            if(s.getUses() == 0)
                auxshield.add(s);
        }
        for(ShieldBooster s: auxshield){
            shieldBoosters.remove(s);
        }
    }
    
    public void discardHangar(){
         hangar = null;
    }
    
    public void discardShieldBooster(int i){
        int size = shieldBoosters.size();
        
        if((i>=0) && (i<size)){
            shieldBoosters.remove(i);
            if(pendingDamage != null){
                pendingDamage.discardShieldBooster();
                cleanPendingDamage();
            }
        }
    }
    
    public void discardShieldBoosterInHangar(int i){
        if (hangar != null && i>=0 && hangar.getShieldBoosters().size() > i){
            hangar.removeShieldBooster(i);
        }
    }
    
    public void discardWeapon(int i){
        int size = weapons.size();
        
        if((i>=0) && (i<size)){
            Weapon w = weapons.remove(i);
            if(pendingDamage != null){
                pendingDamage.discardWeapon(w);
                cleanPendingDamage();
            }
        }
    }
    
    public void discardWeaponInHangar(int i){
        if (hangar != null && i>=0 && hangar.getWeapons().size() > i){
            hangar.removeWeapon(i);
        }
    }
    
    @Override
    public float fire(){
        int size = weapons.size();
        float factor = 1;
        Weapon w;
        
        for(int i=0; i<size; i++){
            w = weapons.get(i);
            factor *= w.useIt();
        }
        
        return ammoPower*factor;     
    }
    
    public float getAmmoPower() {
        return this.ammoPower;
    }
    
    public float getFuelUnits() {
        return this.fuelUnits;
    }
    
    public Hangar getHangar() {
        return hangar;
    }
    
    public String getName() {
        return this.name;
    }
    
    public int getNMedals() {
        return nMedals;
    }
    
    public Damage getPendingDamage() {
        return pendingDamage;
    }
    
    public ArrayList<ShieldBooster> getShieldBoosters() {
        return shieldBoosters;
    }
    
    public float getShieldPower() {
        return shieldPower;
    }
    
    public float getSpeed() {
        float result;
        result = fuelUnits/MAXFUEL;
        
        return result;
    }
    
    public SpaceStationToUI getUIversion() {
        return new SpaceStationToUI(this);
    }
    
    public ArrayList<Weapon> getWeapons() {
        return weapons;
    }
    
    public void mountShieldBooster(int i){
        if(hangar != null){
            if((0<=i) && (i<hangar.getShieldBoosters().size())){
                ShieldBooster shieldBooster = hangar.removeShieldBooster(i);
                if(shieldBooster != null)
                    shieldBoosters.add(shieldBooster);
            }
        }
    }
    
    public void mountWeapon(int i){
        if(hangar != null){
            if((0<=i) && (i<hangar.getWeapons().size())){
                Weapon weapon = hangar.removeWeapon(i);
                if(weapon != null)
                    weapons.add(weapon);
            }
        }
    }
    
    public void move(){
        fuelUnits = fuelUnits - getSpeed();
        
        if(fuelUnits<0){
            fuelUnits = 0;
        }
    }
    
    @Override
    public float protection(){
        int size = shieldBoosters.size();
        float factor = 1;
        ShieldBooster s;
        
        for(int i=0; i<size;i++){
            s = shieldBoosters.get(i);
            factor *= s.useIt();
        }
        
        return factor*shieldPower;
    }
    
    public void receiveHangar(Hangar h){
        if(hangar == null){
            hangar = h; 
        }
    }
    
    public boolean receiveShieldBooster(ShieldBooster s){
        if(hangar != null){
            return hangar.addShieldBooster(s);
        }else return false;
    }
    
    @Override
    public ShotResult receiveShot(float shot){
        float myProtection = protection();
        
        if (myProtection >= shot){
            shieldPower -= SHIELDLOSSPERUNITSHOT*shot;
            shieldPower = Math.max(0.0f,shieldPower);
            return ShotResult.RESIST;
        }else{
            shieldPower = 0.0f;
            return ShotResult.DONOTRESIST;
        } 
    }
    
    public void receiveSupplies(SuppliesPackage s){
        ammoPower = ammoPower + s.getAmmoPower();
        fuelUnits = fuelUnits + s.getFuelUnits();
        if(fuelUnits > MAXFUEL)
            fuelUnits = MAXFUEL;
        shieldPower = shieldPower + s.getShieldPower();
    }
    
    public boolean receiveWeapon(Weapon w){
        if(hangar != null){
            return hangar.addWeapon(w);
        }else return false;
    }
    
    public Transformation setLoot(Loot loot){
        CardDealer dealer = CardDealer.getInstance();
        int h = loot.getNHangars();
        
        if(h>0){
            Hangar hangar = dealer.nextHangar();
            receiveHangar(hangar);
        }
        
        int elements=loot.getNSupplies();
        for(int i=0; i<elements; i++){
            SuppliesPackage sup = dealer.nextSuppliesPackage();
            receiveSupplies(sup);
        }
        
        elements = loot.getNWeapons();
        for(int i=0; i<elements; i++){
            Weapon weap = dealer.nextWeapon();
            receiveWeapon(weap);
        }
        
        elements = loot.getNShields();
        for(int i=0;i<elements; i++){
            ShieldBooster sh = dealer.nextShieldBooster();
            receiveShieldBooster(sh);
        }
        
        int medals = loot.getNMedals();
        
        nMedals += medals;
        
        if(loot.getEfficient()){
            return Transformation.GETEFFICIENT;
        }else{
            if(loot.spaceCity()){
                return Transformation.SPACECITY;
            } else return Transformation.NOTRANSFORM;
        }
    }
    
    public void setPendingDamage(Damage d){
        pendingDamage = d.adjust(weapons, shieldBoosters);
        
        cleanPendingDamage();
    }
    
    public boolean validState() {
        if (pendingDamage == null){
            return true;
        }else return (pendingDamage.hasNoEffect());
    }
     
    @Override
    public String toString(){
        String n = "Nombre: " + name + "\n";
        String p = "AmmoPower: " + ammoPower + "\n";
        String f = "FuelUnits: " + fuelUnits + "\n";    
        String sp = "ShieldPower: " + shieldPower + "\n";
        String m = "NºMedals: " + nMedals + "\n";
        
        String w = "Weapons: ";
        for(Weapon weap : weapons){
            w += weap.toString();
        }
        w += "\n";
        
        String sb = "ShieldBoosters: ";
        for(ShieldBooster shieldb : shieldBoosters){
                sb += shieldb.toString();
        }
        sb += "\n";
        String last = n + p + f + sp + m + w + sb;
        
        String h = "Hangar: ";
        if (hangar != null){
            h += hangar.toString() + "\n";  
        }else h += "null \n";
        last += h;
        
        String pd = "PendingDamage: ";
        if(pendingDamage != null){
             pd += pendingDamage.toString() + "\n"; 
             
        }else pd += "null \n";
        
        last += pd;
        
        return last;
    }
}
